test/test-udp-mmsg.c

```diff
@@ -0,0 +1,124 @@
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy
+ * of this software and associated documentation files (the "Software"), to
+ * deal in the Software without restriction, including without limitation the
+ * rights to use, copy, modify, merge, publish, distribute, sublicense, and/or
+ * sell copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
+ * FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS
+ * IN THE SOFTWARE.
+ */
+
+#include "uv.h"
+#include "task.h"
+
+#include <stdio.h>
+#include <stdlib.h>
+#include <string.h>
+
+#define CHECK_HANDLE(handle) \
+  ASSERT((uv_udp_t*)(handle) == &server || (uv_udp_t*)(handle) == &client)
+
+#define BUFFER_SIZE (64 * 1024)
+#define NUM_SENDS 4
+
+static uv_udp_t server;
+static uv_udp_t client;
+static int sv_recv_cb_called;
+static int close_cb_called;
+static int alloc_cb_called;
+
+
+static void alloc_cb(uv_handle_t* handle,
+                     size_t suggested_size,
+                     uv_buf_t* buf) {
+  CHECK_HANDLE(handle);
+  ASSERT(suggested_size <= BUFFER_SIZE);
+  /* Actually malloc to exercise free'ing the buffer later */
+  buf->base = malloc(BUFFER_SIZE);
+  buf->len = BUFFER_SIZE;
+  alloc_cb_called++;
+}
+
+
+static void close_cb(uv_handle_t* handle) {
+  CHECK_HANDLE(handle);
+  ASSERT(uv_is_closing(handle));
+  close_cb_called++;
+}
+
+
+static void sv_recv_cb(uv_udp_t* handle,
+                       ssize_t nread,
+                       const uv_buf_t* rcvbuf,
+                       const struct sockaddr* addr,
+                       unsigned flags) {
+  ASSERT(nread >= 0);
+
+  if (nread > 0) {
+    ASSERT(nread == 4);
+    ASSERT(addr != NULL);
+    ASSERT(memcmp("PING", rcvbuf->base, nread) == 0);
+
+    sv_recv_cb_called++;
+    if (sv_recv_cb_called == NUM_SENDS) {
+      uv_close((uv_handle_t*)handle, close_cb);
+      uv_close((uv_handle_t*)&client, close_cb);
+    }
+  }
+
+  /* Don't free if the buffer will be reused via mmsg */
+  if (rcvbuf && !(flags & UV_UDP_MMSG_CHUNK))
+    free(rcvbuf->base);
+}
+
+
+TEST_IMPL(udp_mmsg) {
+  struct sockaddr_in addr;
+  uv_buf_t buf;
+  int r;
+  int i;
+
+  ASSERT(0 == uv_ip4_addr("0.0.0.0", TEST_PORT, &addr));
+
+  r = uv_udp_init_ex(uv_default_loop(), &server, AF_UNSPEC | UV_UDP_RECVMMSG);
+  ASSERT(r == 0);
```

Comment:
You probably copied this these another test? Since you're here, can you change them to e.g.:
```suggestion
  ASSERT_EQ(0, uv_udp_init_ex(uv_default_loop(), &server,
                              AF_UNSPEC | UV_UDP_RECVMMSG));
```
That makes debugging tests easier (better comprehensible error messages.)

Reply:
> You probably copied this these another test?

Yep, it's based on [test-udp-try-send.c](https://github.com/libuv/libuv/blob/v1.x/test/test-udp-try-send.c)